packages/sdk/src/contract/contracts/Xtokens/Xtokens.ts

```diff
@@ -1,26 +1,73 @@
 import type { TransactionResponse } from '@ethersproject/abstract-provider';
 import { ContractConfig } from '@moonbeam-network/xcm-builder';
 import { Contract, Signer } from 'ethers';
-import { TransferContractInterface } from '../../contract.interfaces';
+import {
+  WalletClient,
+  WriteContractReturnType,
+  createPublicClient,
+  http,
+} from 'viem';
+import {
+  ContractClient,
+  TransferContractInterface,
+  ViemClient,
+} from '../../contract.interfaces';
+import { isEthersClient, isEthersSigner } from '../../contract.utils';
 import abi from './XtokensABI.json';
 
 export class Xtokens implements TransferContractInterface {
-  readonly address: string = '0x0000000000000000000000000000000000000804';
+  readonly address = '0x0000000000000000000000000000000000000804';
 
   readonly #config: ContractConfig;
 
-  readonly #contract: Contract;
+  readonly #client: ContractClient;
 
-  readonly #signer: Signer;
-
-  constructor(config: ContractConfig, signer: Signer) {
+  constructor(config: ContractConfig, signer: Signer | WalletClient) {
     this.#config = config;
-    this.#signer = signer;
-    this.#contract = new Contract(this.address, abi, signer);
+
+    this.#client = isEthersSigner(signer)
```

Comment:
To simplify stuff I would try to use [getContract](https://viem.sh/docs/contract/getContract.html).